# Machine Learning for 3D Course Capstone

Welcome to the final capstone project of the course.

To complete this course and receive a certification, you will create and host your own Generative 3D model and demo.

### Requirements:

- **Model**: A model hosted under your username, e.g. `hf.co/{username}/{model_name}`.
  - Ensure there is a [model card](https://huggingface.co/docs/hub/en/model-cards).
  - Tag the model with the `image-to-3d` pipeline tag.
  - Apply the correct license if applicable.
  
- **Space**: A space hosted under your username, e.g. `hf.co/spaces/{username}/{space_name}`.
  - Ensure the space converts images (`.png`, `.jpg`) to 3D (`.glb`, `.obj`, `.ply`, `.splat`).
  - Reference your model URL in the space README.

When completed, fill out this [form](https://forms.gle/rQedXFktHPYeikrt6) to receive your certificate.

## What should I do?

This capstone is very open-ended. You can:

- Build something from scratch.
- Train or fine-tune an existing open-source model.
- Clone an existing open-source model and change the demo experience.
- Directly clone an open-source model and demo.

Check out [3D Arena](https://huggingface.co/spaces/dylanebert/3d-arena) for the latest image-to-3D demos to use as starting points.

## But how?

This section will walk you through the easiest way to set up a demo based on [LGM](https://huggingface.co/spaces/ashawkey/LGM), as covered in previous units.

### 1. Duplicate the model.

Start by duplicating the [LGM-full](https://huggingface.co/dylanebert/LGM-full) model.

1. Go to the [Repo duplicator](https://huggingface.co/spaces/huggingface-projects/repo_duplicator).
2. Generate a `write token` in your [account settings](https://hf.co/settings/tokens). Copy it into the `token` field.
3. Enter `dylanebert/LGM-full` in the `source_repo` field.
4. Enter `{username}/{model_name}` in the `dst_repo` field, replacing `{username}` with your Hugging Face account username, and `{model_name}` with any name you like, such as `LGM`.
5. Click Submit.

You will see a link labeled "find your repo **here**", which leads to `http://huggingface.co/{username}/{model_name}`.

Congratulations! You can now use this model with Diffusers, and it will appear in [hf.co/models](https://huggingface.co/models?pipeline_tag=image-to-3d&sort=trending).



### 2. Duplicate the space.

Go to the [LGM Tiny](https://huggingface.co/spaces/dylanebert/LGM-tiny) space, which provides a simplified image-to-3D demo.

1. Click the `Duplicate Space` button.
2. Choose free hardware, or ZeroGPU Nvidia A100 if available.
3. Click `Files` in the top right.
4. Click `app.py` to view the demo source code.
5. Click `edit` to change the code.
6. Replace the two instances of `dylanebert/LGM-full` with your model path, e.g. `{username}/{model_name}`.
7. Click `Commit changes`.

Congratulations! You've created a demo and met the minimum requirements for this capstone project.


## So how do I run it?

The demo requires a GPU, so it won't work on free hardware. However, there are many free options:

1. **Run in this notebook**: Validate the code quickly.
2. **Run locally**: Clone your space and run it locally.
3. **Community grant**: Building something cool? Apply for a community GPU grant in your space settings.
4. **Run via API**: Less flexible, but runs on free hardware.
5. **Just don't**: Alternatively, skip all this and go straight to the [form](https://forms.gle/rQedXFktHPYeikrt6).

### Run in this notebook

Start by installing the necessary dependencies, as in previous sections.

> Note: Ensure the notebook runtime type is set to GPU.

In [1]:
!pip install -r https://huggingface.co/spaces/dylanebert/LGM-tiny/raw/main/requirements.txt

  Using cached tyro-0.8.5-py3-none-any.whl (103 kB)
  Using cached diffusers-0.29.2-py3-none-any.whl (2.2 MB)
  Using cached dearpygui-1.11.1-cp310-cp310-manylinux1_x86_64.whl (2.8 MB)
  Using cached einops-0.8.0-py3-none-any.whl (43 kB)
  Using cached accelerate-0.32.1-py3-none-any.whl (314 kB)
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 12.3/12.3 MB 22.6 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 53.8/53.8 kB 6.3 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 42.5/42.5 kB 2.3 MB/s eta 0:00:00
  Preparing metadata (setup.py) ... done
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 695.9/695.9 kB 27.2 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 3.1/3.1 MB 31.3 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 230.1/230.1 kB 14.9 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 755.5/755.5 MB 1.9 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 6.9/6.9 MB 59.4 MB/s eta 0:00:00
     ━━━━━━━━━━

The code below is exactly the same as in the space `app.py`. To ensure your model is working as expected, replace both instances of `dylanebert/LGM-full` with your `{username}/{model_name}`. Then, run the code.

In [2]:
import shlex
import subprocess

import gradio as gr
import numpy as np
import spaces
import torch
from diffusers import DiffusionPipeline

subprocess.run(
    shlex.split(
        "pip install https://huggingface.co/spaces/dylanebert/LGM-mini/resolve/main/wheel/diff_gaussian_rasterization-0.0.0-cp310-cp310-linux_x86_64.whl"
    )
)

pipeline = DiffusionPipeline.from_pretrained(
    "dylanebert/LGM-full",
    custom_pipeline="dylanebert/LGM-full",
    torch_dtype=torch.float16,
    trust_remote_code=True,
).to("cuda")


@spaces.GPU
def run(image):
    input_image = np.array(image, dtype=np.float32) / 255.0
    splat = pipeline(
        "", input_image, guidance_scale=5, num_inference_steps=30, elevation=0
    )
    splat_file = "/tmp/output.ply"
    pipeline.save_ply(splat, splat_file)
    return splat_file


demo = gr.Interface(
    fn=run,
    title="LGM Tiny",
    description="An extremely simplified version of [LGM](https://huggingface.co/ashawkey/LGM). Intended as resource for the [ML for 3D Course](https://huggingface.co/learn/ml-for-3d-course/unit0/introduction).",
    inputs="image",
    outputs=gr.Model3D(),
    examples=[
        "https://huggingface.co/datasets/dylanebert/iso3d/resolve/main/jpg@512/a_cat_statue.jpg"
    ],
    cache_examples=True,
    allow_duplication=True,
)
demo.queue().launch()

The cache for model files in Transformers v4.22.0 has been updated. Migrating your old cache. This is a one-time only operation. You can interrupt this and resume the migration later on by calling `transformers.utils.move_cache()`.


0it [00:00, ?it/s]

/usr/local/lib/python3.10/dist-packages/huggingface_hub/utils/_token.py:89: UserWarning: 
The secret `HF_TOKEN` does not exist in your Colab secrets.
To authenticate with the Hugging Face Hub, create a token in your settings tab (https://huggingface.co/settings/tokens), set it as secret in your Google Colab and restart your session.
You will be able to reuse this secret in all of your notebooks.
Please note that authentication is recommended but still optional to access public models or datasets.
  warnings.warn(


model_index.json:   0%|          | 0.00/511 [00:00<?, ?B/s]

pipeline.py:   0%|          | 0.00/59.6k [00:00<?, ?B/s]

Fetching 20 files:   0%|          | 0/20 [00:00<?, ?it/s]

(…)ature_extractor/preprocessor_config.json:   0%|          | 0.00/546 [00:00<?, ?B/s]

image_encoder/config.json:   0%|          | 0.00/563 [00:00<?, ?B/s]

text_encoder/config.json:   0%|          | 0.00/610 [00:00<?, ?B/s]

lgm/lgm.py:   0%|          | 0.00/25.0k [00:00<?, ?B/s]

scheduler/scheduler_config.json:   0%|          | 0.00/502 [00:00<?, ?B/s]

lgm/config.json:   0%|          | 0.00/65.0 [00:00<?, ?B/s]

model.safetensors:   0%|          | 0.00/1.26G [00:00<?, ?B/s]

tokenizer/special_tokens_map.json:   0%|          | 0.00/460 [00:00<?, ?B/s]

tokenizer/tokenizer_config.json:   0%|          | 0.00/855 [00:00<?, ?B/s]

tokenizer/merges.txt:   0%|          | 0.00/525k [00:00<?, ?B/s]

diffusion_pytorch_model.safetensors:   0%|          | 0.00/830M [00:00<?, ?B/s]

unet/config.json:   0%|          | 0.00/403 [00:00<?, ?B/s]

tokenizer/vocab.json:   0%|          | 0.00/1.06M [00:00<?, ?B/s]

model.safetensors:   0%|          | 0.00/681M [00:00<?, ?B/s]

vae/config.json:   0%|          | 0.00/601 [00:00<?, ?B/s]

unet/mv_unet.py:   0%|          | 0.00/34.4k [00:00<?, ?B/s]

diffusion_pytorch_model.safetensors:   0%|          | 0.00/167M [00:00<?, ?B/s]

diffusion_pytorch_model.safetensors:   0%|          | 0.00/1.88G [00:00<?, ?B/s]

Keyword arguments {'trust_remote_code': True} are not expected by LGMFullPipeline and will be ignored.


Loading pipeline components...:   0%|          | 0/8 [00:00<?, ?it/s]

/root/.cache/huggingface/modules/diffusers_modules/local/lgm.py:334: UserWarning: xFormers is available (Attention)
  warnings.warn("xFormers is available (Attention)")


Caching examples at: '/content/gradio_cached_examples/15'
Caching example 1/1


  0%|          | 0/30 [00:00<?, ?it/s]

Setting queue=True in a Colab notebook requires sharing enabled. Setting `share=True` (you can turn this off by setting `share=False` in `launch()` explicitly).

Colab notebook detected. To show errors in colab notebook, set debug=True in launch()
Running on public URL: https://2b664f0df7900fd8c5.gradio.live

This share link expires in 72 hours. For free permanent hosting and GPU upgrades, run `gradio deploy` from Terminal to deploy to Spaces (https://huggingface.co/spaces)


### Run locally

The instructions below are tested on an RTX 4090 on WSL2 Ubuntu 22.04. Instructions will differ and may not work, depending on your setup.

1. Install `git`, `python 3.10`, and `cuda` if not already installed.
2. Open your terminal.
3. Clone your space repository, replacing the URL below with your space URL.

In [ ]:
git clone https://huggingface.co/spaces/dylanebert/LGM-tiny

4. Navigate into the space folder.

In [ ]:
cd LGM-tiny

5. Create a virtual environment and install necessary dependencies.

In [ ]:
python -m venv venv
source venv/bin/activate
pip install -r requirements.txt

6. Run the demo.

In [ ]:
python app.py

### Use via API



To run via API, instead of duplicating the [LGM-tiny](https://huggingface.co/spaces/dylanebert/LGM-tiny) space, duplicate the [LGM-tiny-api](https://huggingface.co/spaces/dylanebert/LGM-tiny-api) space. This contains the following `app.py`.

In [3]:
import gradio as gr
from gradio_client import Client, file


def run(image_url):
    client = Client("dylanebert/LGM-tiny")
    image = file(image_url)
    result = client.predict(image, api_name="/predict")
    return result


demo = gr.Interface(
    fn=run,
    title="LGM Tiny API",
    description="An API wrapper for [LGM Tiny](https://huggingface.co/spaces/dylanebert/LGM-tiny). Intended as a resource for the [ML for 3D Course](https://huggingface.co/learn/ml-for-3d-course).",
    inputs=gr.Textbox(label="Image URL", placeholder="Enter image URL, e.g. https://huggingface.co/datasets/dylanebert/iso3d/resolve/main/jpg@512/a_cat_statue.jpg"),
    outputs=gr.Model3D(),
    examples=[
        "https://huggingface.co/datasets/dylanebert/iso3d/resolve/main/jpg@512/a_cat_statue.jpg"
    ],
    allow_duplication=True,
)
demo.queue().launch()

Setting queue=True in a Colab notebook requires sharing enabled. Setting `share=True` (you can turn this off by setting `share=False` in `launch()` explicitly).

Colab notebook detected. To show errors in colab notebook, set debug=True in launch()
Running on public URL: https://f30947e1f9c17f5b83.gradio.live

This share link expires in 72 hours. For free permanent hosting and GPU upgrades, run `gradio deploy` from Terminal to deploy to Spaces (https://huggingface.co/spaces)


This will work on CPU, but relies on the original LGM-tiny, instead of your custom model. However, is your focus is on UI/UX or downstream tasks, this may be acceptable.

## Conclusion

Thank you for participating in the Machine Learning for 3D Course!

Join the [discord](https://hf.co/join/discord) and share your thoughts, questions, and capstone projects in the #3d channel.